src/Altinn-application-owner-system/Functions/Services/Implementation/AuthenticationService.cs

```diff
@@ -0,0 +1,157 @@
+            IMaskinPortenClientWrapper maskinPortenClientWrapper,
+            IAuthenticationClientWrapper authenticationClientWrapper,
+            ILogger<AuthenticationService> logger)
+        {
+            _keyVaultSettings = keyVaultSettings.Value;
+            _altinnIntegratorSettings = altinnIntegratorSettings.Value;
+            _keyVaultService = keyVaultService;
+            _maskinportenClientWrapper = maskinPortenClientWrapper;
+            _authenticationClientWrapper = authenticationClientWrapper;
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Creates a altinn token. First login to MaskinPorten and then call altinn to convert to Altinn token.
+        /// </summary>
+        /// <returns>The altinn token</returns>
+        public async Task<string> GetAltinnToken()
+        {
+            // To reduce traffic we cache this.
+            if (!string.IsNullOrEmpty(_altinnToken) && _tokenTimeout > DateTime.Now)
+            {
+                return _altinnToken;
+            }
+
+            string jwtAssertion = await GetJwtAssertion();
+
+            FormUrlEncodedContent content = GetUrlEncodedContent(jwtAssertion);
+            string maskinPortenToken = await _maskinportenClientWrapper.PostToken(content);
+            
+            if (!string.IsNullOrEmpty(maskinPortenToken))
+            {
+                var accessTokenObject = JsonConvert.DeserializeObject<JObject>(maskinPortenToken);
+
+                string altinnToken = await _authenticationClientWrapper.ConvertToken(accessTokenObject.GetValue("access_token").ToString());
+
+                _altinnToken = altinnToken;
+                _tokenTimeout = DateTime.Now.AddSeconds(9);
+                return altinnToken;
+            }
+
+            return null;
+        }
+
+        private FormUrlEncodedContent GetUrlEncodedContent(string assertion)
+        {
+            FormUrlEncodedContent formContent = new FormUrlEncodedContent(new List<KeyValuePair<string, string>>
+            {
+                new KeyValuePair<string, string>("grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer"),
+                new KeyValuePair<string, string>("assertion", assertion),
+            });
+
+            return formContent;
+        }
+
+        /// <summary>
+        /// Creates the JWT Assertion used to authenticate system in Maskinporten
+        /// </summary>
+        /// <returns></returns>
+        public async Task<string> GetJwtAssertion()
+        {
+            DateTimeOffset dateTimeOffset = new DateTimeOffset(DateTime.UtcNow);
+            Guid clientId = new Guid(_altinnIntegratorSettings.MaskinPortenClientId);
+
+            X509Certificate2 cert = null;
+            if (!string.IsNullOrEmpty(_altinnIntegratorSettings.LocalCertThumbprint))
+            {
+               cert = GetCertificateFromKeyStore("111bd85e12ef3dce6ec7b2c0011d4e9cbc4f896f", StoreName.My, StoreLocation.LocalMachine);
```

Comment:
"111bd85e12ef3dce6ec7b2c0011d4e9cbc4f896f" should probably be _altinnIntegratorSettings.LocalCertThumbprint?